rfcs/20201119-change-memory-init-behavior/README.md

```diff
@@ -0,0 +1,272 @@
+> situation happens if memory objects are kept alive and reused for
+> example. Such a flag would help if we have situation where doing
+> tail handling is faster than writing zeros to padding.
+
+### Option 2. Move destination zero-padding to primitive execution, and expose a memory initialization primitive
+
+Here it would be essentially the same as option 1. However, to
+simplify the transition for users that have custom computations on
+blocked buffers, we would expose a new primitive to initialize memory.
+There are two sub-options here:
+- option 2.a: introduce an "inplace" reorder. If the user calls
+  reorder with the same input and output, the only effect that should
+  be expected is explicit zero padding. The bad side of this option is
+  that it is somehow a "trick" since this reorder would break
+  assumption a. But if this is aiming only advanced users writing
+  kernels that compute on blocked buffers, then maybe we can live with
+  this docmented exception.
+- option 2.b: introduce a new `memory_cleanup` primitive, that should
+  be used after every custom computation done on a oneDNN memory
+  handle. This one has the merit of being more explicit than calling a
+  reorder.
+
+Both approaches try to simplify explicit zero padding from users that
+write custom kernels. However, it is unclear if they would indeed be
+useful, since they require that the component responsible for the
+custom kernel call also has access to the dnnl memory object. This is
+the case for iDeep, but it is not clear if other potential users are
+in the same situation.
+
+This achieves goals 1 and 2, but not 3. However, it provides a simple
+mitigation for goal 3 that should not impact performance too much
+(should still be better than what we currently have).
+
+## Option 3. Move source zero-padding to primitive execution
+
+Here we would change the assumptions that each primitive
+implementation can make:
+a. The inputs of a primitive are not always properly zero padded
+b. The outputs of a primitive do not have to be properly zero-padded.
+
+This can reduce the number of calls to zero padding in two ways:
+- primitives can skip zero padding on their source if they support
+  tail handling
+- when the users pass handles between two primitive, it will result in
+  one zero pad call instead of two.
+
+Achieves goal 2 and 3. However, goal 1 is only partially resolved
+since this mechanism avoids some explicit zero-padding but not all. In
+particular, we do not have many implementations that support tail
+computation, so removing the zero padding overhead completely could
+require a large implementation effort.
+
+## Option 4. Move zero-padding to primitive execution, and add metadata to avoid spurious zero padding.
+This option uses no assumption on the zero-padding of its input/output
+tensors. Here, we would carry this information through metadata in the
+memory descriptors. That metadata can then be used by primitive
```

Comment:
I think we should not add the state itself to memory descriptor. Memory descriptor is conceptually constant except for a case when created with any. Also memory descriptor defines data representation, it doesn't depend on a buffer or it's content. I would prefer to add a flag to memory descriptor that there is an additional space in buffer with meta data. Any implementation can look onto this flag and if it is set then primitive can work with metadata.

Reply:
Just to clarify: the metadata is constant here (non mutable). Maybe metadata was the wrong wording.
When a user calls `memory_desc_init`, we will set the metadata to `unknown`.
When a primitive is queried for a memory descriptor, the state will be either clean if it accepts only clean memories, or dirty otherwise (but never `unknown`).

If the user propagates the layouts properly through memory descriptors, then the metadata of the destination memory produced by a primitive will be carried forward to the next primitive input. Now when the memory descriptors get compared (p0.dst_desc() != p1.src_desc()) for reorder we would use the following truth table:
- dirty == dirty -> true
- clean == clean -> true
- unknown == unknown -> true
- unknown == dirty -> true
- unknown == clean -> false
- clean == dirty -> false
 

We thought of putting dynamic (mutable) metadata. However, it could not go in the memory descriptor as you say, because the memory descriptor should be a constant data. 
Because framework integration passes handles, the other option would be to put the metadata in the handle. In that case, we could have issues in two ways:
- that will complicate the code of users that have custom kernels working on oneDNN memory buffers.
- that will bring some overhead in case the handle is a GPU memory (we might have to move this metadata to the host).